src/Abstractions/TaskName.cs

```diff
@@ -8,127 +8,169 @@ namespace Microsoft.DurableTask;
+  /// <param name="b">The second <see cref="TaskName"/> to compare.</param>
+  /// <returns><c>true</c> if the two <see cref="TaskName"/> objects are not equal; otherwise <c>false</c>.</returns>
+  public static bool operator !=(TaskName a, TaskName b)
+  {
+    return !a.Equals(b);
+  }
+
+  /// <summary>
+  /// Gets a value indicating whether to <see cref="TaskName"/> objects
+  /// are equal using value semantics.
+  /// </summary>
+  /// <param name="other">The other object to compare to.</param>
+  /// <returns><c>true</c> if the two objects are equal using value semantics; otherwise <c>false</c>.</returns>
+  public bool Equals(TaskName other)
+  {
+    return string.Equals(this.Name, other.Name, StringComparison.OrdinalIgnoreCase)
+      && string.Equals(this.Version, other.Version, StringComparison.OrdinalIgnoreCase);
+  }
+
+  /// <summary>
+  /// Gets a value indicating whether to <see cref="TaskName"/> objects
+  /// are equal using value semantics.
+  /// </summary>
+  /// <param name="obj">The other object to compare to.</param>
+  /// <returns><c>true</c> if the two objects are equal using value semantics; otherwise <c>false</c>.</returns>
+  public override bool Equals(object? obj)
+  {
+    if (obj is not TaskName other)
+    {
+      return false;
     }
 
-    /// <summary>
-    /// Calculates a hash code value for the current <see cref="TaskName"/> instance.
-    /// </summary>
-    /// <returns>A 32-bit hash code value.</returns>
-    public override int GetHashCode()
+    return this.Equals(other);
+  }
+
+  /// <summary>
+  /// Calculates a hash code value for the current <see cref="TaskName"/> instance.
+  /// </summary>
+  /// <returns>A 32-bit hash code value.</returns>
+  public override int GetHashCode()
+  {
+    if (string.IsNullOrEmpty(this.Version))
     {
-        return StringComparer.OrdinalIgnoreCase.GetHashCode(this.Name);
+      return StringComparer.OrdinalIgnoreCase.GetHashCode(this.Name);
     }
 
-    /// <summary>
-    /// Gets the string value of the current <see cref="TaskName"/> instance.
-    /// </summary>
-    /// <returns>The name and optional version of the current <see cref="TaskName"/> instance.</returns>
-    public override string ToString()
+    return StringComparer.OrdinalIgnoreCase.GetHashCode(this.Name) * 31
+      + StringComparer.OrdinalIgnoreCase.GetHashCode(this.Version);
+  }
+
+  /// <summary>
+  /// Gets the string value of the current <see cref="TaskName"/> instance.
+  /// </summary>
+  /// <returns>The name and optional version of the current <see cref="TaskName"/> instance.</returns>
+  public override string ToString()
+  {
+    if (string.IsNullOrEmpty(this.Version))
+    {
+      return this.Name;
+    }
+    else
     {
-        if (string.IsNullOrEmpty(this.Version))
-        {
-            return this.Name;
-        }
-        else
-        {
-            return this.Name + ":" + this.Version;
-        }
+      return this.Name + ":" + this.Version;
```

Comment:
We should have a discussion on what separator to use. I like `@`, as that is what we have used for distributed tracing to indicate the version - and ":" was used to separate the task type: `{task_type}:{name}@{version}`

However, we already use `@` to separate entity name and key - will that be an issue? Or is that a good thing to use `@` as a separator in both for consistency.

Reply:
I think using `@` could break entity ID parsing. Looking at [this code](https://github.com/Azure/durabletask/blob/d4b98d46158e470c6b769601cd76a86d9bb6d096/src/DurableTask.Core/Entities/EntityId.cs#L54-L78), I can see that entity IDs take the form of `{name}@{key}` and the code that parses this looks for the first instance of `@` to separate the name from the key. If we inject a version into the name, we make it `{name}@{version}@{key}`, and the key for entities will be parsed as `{version}@{key}` instead of `{key}`.

Reply:
@cgillum entity Ids is the _instance_ ID - this is the _name_. They are not the same field.

Reply:
@cgillum yeah I am trying to think through how that behaves. One thing is that it is a separate type for entities. We have a few options:

1. use `@` for separator. Entity instance IDs can then look like `@{entity}@{version}@{key}` (or we can do key first, then version). **RISK**: keys could already contain `@`. This could break those instances.
    - @sebastianburckhardt suggested we could prefix with double `@@` to indicate this entity has a version field. So `@myEntity@some@key` -> no version field. `@@myEntity@version@some@key` -> version field present.
3. use a different character. `@{entity}?{version}@{key}` (or key first). **RISK**: this character could already be used by customers in name _or_ key
4. Alternatively, we could opt to _not_ support versioning of entities just yet

Reply:
Want to clarify that I am not opposed to a different separator char for version. We haven't officially GA'd distributed tracing v2, so we could update that spec with the new separator. But what one exactly?

Some that I have ruled out myself in the past:

- `/` - avoided this so it wasn't confused with HTTP routes in distributed tracing
- `%` - often used for encoding special chars, so wanted to avoid that
- `*` - wild card, wanted to avoid that.
- `!` - used in netherite for partition targeting. We are looking to also support that in other backends.
- `:` - used in distributed tracing to separate the task-type. ie: `orchestration:name@version` or `activity:name@version`

Reply:
Can we merge this PR first and then open a new issue to trace and update the separator after we have a final decision? 

Because today is my last day, I'm afraid that I can't continue to update the code in this PR because of permissions.

Reply:
@cgillum @jviau please have a look at this PR and let's make the final decision.